simplesat/dependency_solver.py

```diff
@@ -78,40 +79,50 @@ def _create_rules(self, request):
         return requirement_ids, list(rules_generator.iter_rules())
 
 
-def _connected_packages(solution, pkg_ids, pool):
-    """ Return packages in `solution` which are associated with `pkg_ids`. """
+def _connected_packages(solution, root_ids, pool):
+    """ Return packages in `solution` which are associated with `root_ids`. """
 
     # Our strategy is as follows:
     # ... -> pkg.dependencies -> pkg strings -> ids -> _id_to_package -> ...
 
-    # We only need to identify packages which will be installed
-    sol_set = set(s for s in solution if s > 0)
+    def get_name(pkg_id):
+        return pool._id_to_package[abs(pkg_id)].name
+
+    root_names = {get_name(pkg_id) for pkg_id in root_ids}
 
-    # This dict can recover ids from the strings produced by pkg.dependencies
-    package_string_to_id = {}
-    for pkg_id in sol_set:
-        pkg = pool._id_to_package[pkg_id]
-        pkg_key = pkg.name
-        package_string_to_id[pkg_key] = pkg_id
+    solution_name_to_id = {
+        get_name(pkg_id): pkg_id for pkg_id in solution
+        if pkg_id > 0
+    }
+
+    solution_root_ids = set(
+        pkg_id for name, pkg_id in solution_name_to_id.items()
+        if name in root_names
+    )
 
     def neighborfunc(pkg_id):
-        """ Given a pkg id, return the pkg ids of the dependencies that
-        appeared in our solution. """
+        """ Given a pkg id, return the pkg ids of the immediate dependencies
+        that appeared in our solution. """
         dep_strings = pool._id_to_package[pkg_id].dependencies
-        pkg_keys = (
+        pkg_names = (
             Requirement.from_legacy_requirement_string(d).name
             for d in dep_strings
         )
-        neighbors = set(package_string_to_id[key] for key in pkg_keys)
+        neighbors = set(solution_name_to_id[name] for name in pkg_names)
         return neighbors
 
     # Each package can root its own independent graph, so we must start at
     # each one individually
-    connected = set()
-    for pkg_id in pkg_ids:
-        # We pass in `connected` to avoid re-walking a graph we've seen before
-        connected.update(_connected_nodes(pkg_id, neighborfunc, connected))
-
+    should_include = set()
+    for pkg_id in solution_root_ids:
+        # We pass in `should_keep` to avoid re-walking a subgraph
+        nodes = _connected_nodes(pkg_id, neighborfunc, should_include)
+        should_include.update(nodes)
+    assert should_include.issuperset(solution_root_ids)
```

Comment:
This `assert` isn't bringing much value, the solution will always contain the input by construction, however it's very cheap so let's leave it for now.